AstSemantics.md

```diff
@@ -273,39 +275,53 @@ nested. This guarantees that all resulting control flow graphs are well-structur
 
 ## Calls
 
-Direct calls to a function specify the callee by index into a function table.
+Each function has a *signature*, which consists of:
 
-  * `call_direct`: call function directly
+  * Return types, which are a sequence of local types
+  * Argument types, which are a sequence of local types
+
+Note that WebAssembly itself does not support variable-length argument lists.
+C and C++ compilers are expected to implement this functionality by storing
+an array of arguments in a buffer in linear memory and passing a pointer to
+the buffer.
 
-Each function has a signature in terms of expression types, and calls must match
-the function signature
-exactly. [Imported functions](MVP.md#code-loading-and-imports) also have
-signatures and are added to the same function table and are thus also callable
-via `call_direct`.
+In the MVP, the length of the return types vector may only be 0 or 1. This
+restriction may be lifted in the future with the addition of support for
+[multiple return values](FutureFeatures.md#multiple-return-values").
 
-Indirect calls may be made to a value of function-pointer type. A function-
-pointer value may be obtained for a given function as specified by its index
-in the function table.
+There are two forms of calls:
 
+  * `call_direct`: call function directly
   * `call_indirect`: call function indirectly
-  * `addressof`: obtain a function pointer value for a given function
 
-Function-pointer values are comparable for equality and the `addressof` operator
-is monomorphic. Function-pointer values can be explicitly coerced to and from
-integers (which, in particular, is necessary when loading/storing to memory
-since memory only provides integer types). For security and safety reasons,
-the integer value of a coerced function-pointer value is an abstract index and
-does not reveal the actual machine code address of the target function.
+Direct calls identify their function statically. Indirect calls have a
+funcptr operand which identifies the function at runtime.
+
+Calls have a signature, which is the expected return types and argument types
+(ignoring the funcptr operand, in the case of `call_indirect`) of the
+AST node. Call operations trap if the signature of the call differs from the
+signature of the called function.
 
-In the MVP, function pointer values are local to a single module. The
+### Function pointers
+
+Function pointer values are obtained through the use of a special operator:
+
+  * `addressof`: obtain a funcptr value for a given statically-identified function
+
+and are comparable for equality:
+
+  * `funcptr.eq`: funcptr compare equal
+
+Note that it is not possible to directly observe the bits of a funcptr
+value. They may be [converted into integers][], but the integers only hold an
+index into the *function table*, a table with an entry for each function
+appended to the table in the order that they are loaded into the program.
```

Comment:
On Thu, Jul 23, 2015 at 8:43 PM, Dan Gohman <notifications@github.com>
wrote:

> In AstSemantics.md
> <https://github.com/WebAssembly/design/pull/278#discussion_r35356329>:
>
> >
> > -In the MVP, function pointer values are local to a single module. The
> > +### Function pointers
> > +
> > +Function pointer values are obtained through the use of a special operator:
> > +
> > +  * `addressof`: obtain a funcptr value for a given statically-identified function
> > +
> > +and are comparable for equality:
> > +
> > +  * `funcptr.eq`: funcptr compare equal
> > +
> > +Note that it is not possible to directly observe the bits of a funcptr
> > +value. They may be [converted into integers][], but the integers only hold an
> > +index into the *function table*, a table with an entry for each function
> > +appended to the table in the order that they are loaded into the program.
>
> There's a basic design decision to be made somewhere about whether we
accept a dynamic signature check at calls/decodes or whether we try to
design a static mechanism. The addition of local types (that include
signatures) for function pointers only makes sense if we can remove the
need for most, if not all dynamic checks.


> Per the discussions in #89
> <https://github.com/WebAssembly/design/issues/89>, there's one table, not
> one table per signature, and:
>
> Call operations trap if the signature of the call differs from the
> signature of the called function
>
>
That implies a dynamic signature check, which I thought this proposal was
trying to avoid.

>  The order is deterministic if the application loads functions into the
> program in a deterministic fashion:
>
> appended to the table in the order that they are loaded into the program
>
> Also, I don't expect this PR is the last word on function pointers. This
> is just trying to update the text to where the current discussions have
> lead and clean up the text to facilitate the next rounds of discussion.
>
The idea of using multiple function tables, with one per signature, is to
avoid the need for a dynamic signature check and also the need for function
types to be local types, since the program only ever manipulates indices.
To be clear, I am not sure if that's a better solution, since it will
complicate dynamic linking, but it does have those advantages.

A key issue really is how C++ will encode vtables into wasm. Per-signature
function tables can work in an mtable scenario, where the vtable is
represented as a class number and the class number is used to index into
per-virtual-method tables. That requires numbering the classes in a
specific manner, but is dense and single-load efficient.


> —
> Reply to this email directly or view it on GitHub
> <https://github.com/WebAssembly/design/pull/278/files#r35356329>.
>